netket/utils/group/point_group.py

```diff
@@ -0,0 +1,581 @@
+
+    Internally, the transformation is stored as the affine matrix [[W,w],[0,1]],
+    such that matrix inverses and multiplication correspond to inverting and
+    multiplying the transformation itself.
+    """
+
+    _affine: Array
+    """
+    A 2D array specifying the affine transformation. It has to be of the block
+    structure [[W,w],[0,1]].
+    """
+
+    def __pre_init__(self, W: Array, w: Optional[Array] = None) -> Tuple[Tuple, Dict]:
+        W = np.asarray(W)
+        if W.ndim != 2:
+            raise ValueError("W must be a 2D matrix")
+        if W.shape[0] != W.shape[1]:
+            raise ValueError("W must be a square matrix")
+        ndim = W.shape[0]
+        if not np.allclose(W.T @ W, np.eye(ndim)):
+            raise ValueError("W must be an orthogonal matrix")
+        if w is None:
+            w = np.zeros((ndim, 1))
+        else:
+            w = np.asarray(w).reshape(ndim, 1)
+        return ((np.block([[W, w], [np.zeros(ndim), 1]]),), dict())
+
+    @property
+    def affine_matrix(self) -> Array:
+        return self._affine
+
+    @property
+    def matrix(self) -> Array:
+        return self._affine[:-1, :-1]
+
+    @property
+    def translation(self) -> Array:
+        return self._affine[:-1, -1]
+
+    @property
+    def ndim(self) -> int:
+        return self._affine.shape[0] - 1
+
+    @property
+    def is_proper(self) -> bool:
+        return np.isclose(np.linalg.det(self.matrix), 1.0)
+
+    @property
+    def is_symmorphic(self) -> bool:
+        return np.allclose(self.translation, 0.0)
+
+    def __call__(self, x):
+        return np.tensordot(x, self.matrix.T, axes=1) + self.translation
+
+    def preimage(self, x):
+        return np.tensordot(x - self.translation, self.matrix, axes=1)
+
+    def k_action(self, x):
+        # wave vectors are not translated
+        return np.tensordot(x, self.matrix.T, axes=1)
+
+    def __hash__(self):
+        return hash(HashableArray(comparable(self._affine)))
+
+    def __eq__(self, other):
+        if isinstance(other, PGSymmetry):
+            return HashableArray(comparable(self._affine)) == HashableArray(
+                comparable(other._affine)
+            )
+        else:
+            return False
+
+    def change_origin(self, origin: Array) -> "PGSymmetry":
+        """Returns a `PGSymmetry` representing a pure point-group transformation
+        around `origin` with transformation matrix `self._W`."""
+        return PGSymmetry(
+            self.matrix, (np.eye(self.ndim) - self.matrix) @ np.asarray(origin)
+        )
+
+    @struct.property_cached
+    def _name(self) -> str:
+        if self.ndim == 2:
+            return _2D_name(self.matrix, self.translation)
+        elif self.ndim == 3:
+            return _3D_name(self.matrix, self.translation)
+        else:
+            return f"PGSymmetry({self.matrix}, {self.translation})"
+
+    def __repr__(self):
+        return self._name
+
+    def __array__(self, dtype: DType = None):
```

Comment:
Is this actually used anywhere?
Because the matrix is an N+1 x N+1 matrix, right?
so you can do `PGGroup @ ket` but you cannot do `np.asarray(PGGroup) @ ket`.

Reply:
It is used in `PointGroup` to help build product tables etc. We can rename it `affine_matrix` or something if `np.asarray(PGGroup) @ ket` being weird is a concern.

Reply:
I'd be in favour of this and remove `np.asarray` for this

Reply:
Turns out `affine_matrix` already exists!